File: github.com/functional-programmin/Cars.java
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Cars {

    public static void main(String[] args) throws InterruptedException {
        State state = new State();
        new Cars().race(state);
    }

    private void race(State state) throws InterruptedException {
        draw(state);
        int maxSteps = state.carPositions.stream()
                .reduce(1, (max, item) -> item > max ? item : max);
        if (maxSteps < state.steps) {
            race(runStepOfRace(state));
        }
    }

    private void draw(State state) {
        System.out.println();
        IntStream.range(0, state.carNames.size())
                .mapToObj(i -> outputCar(state.carNames.get(i), state.carPositions.get(i)))
                .forEach(System.out::println);
        System.out.println();
    }

    private String outputCar(String carName, int carPosition) {
        String distance = String.join("", Collections.nCopies(carPosition, "-"));
        return String.format("%-12s %-7s|", carName, distance);
    }

    private State runStepOfRace(State state) throws InterruptedException {
        Thread.sleep(state.stepTime * 1000);
        State newState = new State();
        newState.steps = state.steps;
        newState.stepTime = state.stepTime;
        newState.carPositions = moveCars(state.carPositions);
        newState.carNames = state.carNames;
        return newState;
    }

    private List<Integer> moveCars(List<Integer> carPositions) {
        return carPositions.stream()
                .map(x -> new Random().nextInt(10) > 3 ? x + 1 : x)
                .collect(Collectors.toList());
    }


    private static class State {
        int steps = 7;
        int stepTime = 3;
        List<Integer> carPositions = Arrays.asList(1, 1, 1);
        List<String> carNames = Arrays.asList("batmobile", "meteor", "turtle van");
    }
}

File: cars.py
import time
from functools import reduce
from random import random


def move_cars(car_positions):
    return list(map(lambda x: x + 1 if random() > 0.3 else x,
                car_positions))


def output_car(car_name, car_position):
    return f'{car_name.ljust(12)} {("-" * car_position).ljust(7)}|'


def run_step_of_race(state):
    time.sleep(state['step_time'])
    return {'steps': state['steps'],
            'step_time': state['step_time'],
            'car_positions': move_cars(state['car_positions']),
            'car_names': state['car_names']}


def draw(state):
    print()
    print('\n'.join(map(output_car, state['car_names'], state['car_positions'])), end='')
    print()


def race(state):
    draw(state)
    if reduce(lambda start, item: item if item > start else start,
              state['car_positions']) < state['steps']:
        race(run_step_of_race(state))


race({'steps': 7,
      'step_time': 3,
      'car_positions': [1, 1, 1],
      'car_names': ['batmobile', 'meteor', 'turtle van']})

